src/OpenTelemetry/Metrics/Tags.cs

```diff
@@ -52,33 +34,102 @@ public Tags(KeyValuePair<string, object?>[] keyValuePairs)
 
     public readonly bool Equals(Tags other)
     {
-        var length = this.KeyValuePairs.Length;
+        var ourKvps = this.KeyValuePairs;
+        var theirKvps = other.KeyValuePairs;
 
-        if (length != other.KeyValuePairs.Length)
+        if (ourKvps.Length != theirKvps.Length)
         {
             return false;
         }
 
-        for (int i = 0; i < length; i++)
+#if NET6_0_OR_GREATER
+        // Note: This loop uses unsafe code (pointers) to elide bounds checks on
+        // two arrays we know to be of equal length.
+        var cursor = ourKvps.Length;
+        if (cursor > 0)
+        {
+            ref var ours = ref MemoryMarshal.GetArrayDataReference(ourKvps);
+            ref var theirs = ref MemoryMarshal.GetArrayDataReference(theirKvps);
+            while (true)
+            {
+                // Equality check for Keys
+                if (!ours.Key.Equals(theirs.Key))
+                {
+                    return false;
+                }
+
+                // Equality check for Values
+                if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
+                {
+                    return false;
+                }
+
+                if (--cursor == 0)
+                {
+                    break;
+                }
+
+                ours = ref Unsafe.Add(ref ours, 1);
+                theirs = ref Unsafe.Add(ref theirs, 1);
+            }
+        }
+#else
+        for (int i = 0; i < ourKvps.Length; i++)
         {
-            ref var left = ref this.KeyValuePairs[i];
-            ref var right = ref other.KeyValuePairs[i];
+            ref var ours = ref ourKvps[i];
+
+            // Note: Bounds check happens here for theirKvps element access
+            ref var theirs = ref theirKvps[i];
 
             // Equality check for Keys
-            if (!left.Key.Equals(right.Key, StringComparison.Ordinal))
+            if (!ours.Key.Equals(theirs.Key))
             {
                 return false;
             }
 
             // Equality check for Values
-            if (!left.Value?.Equals(right.Value) ?? right.Value != null)
+            if (!ours.Value?.Equals(theirs.Value) ?? theirs.Value != null)
             {
                 return false;
             }
         }
+#endif
 
         return true;
     }
 
     public override readonly int GetHashCode() => this.hashCode;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static int ComputeHashCode(KeyValuePair<string, object?>[] keyValuePairs)
```

Comment:
Here is a diff to show the changes side-by-side...

```diff
private static int ComputeHashCode(KeyValuePair<string, object?>[] keyValuePairs)
{
        this.KeyValuePairs = keyValuePairs;

-       for (int i = 0; i < this.KeyValuePairs.Length; i++)
        for (int i = 0; i < keyValuePairs.Length; i++)
        {
-            ref var item = ref this.KeyValuePairs[i];
+            ref var item = ref keyValuePairss[i];

-            hashCode.Add(item.Key));
+            hashCode.Add(item.Key.GetHashCode()));

            hashCode.Add(item.Value);
        }
}
```

* The first change is to use the local `keyValuePairs` and not property `this.KeyValuePairs` for array access. Turns out the JIT won't elide bounds checks for array access unless it can tell for sure the reference to the array won't change during the loop. That's what using the local does.

* The second change is to call `item.Key.GetHashCode()` directly and pass in the result. This elides the `null` check in [HashCode.Add](https://github.com/dotnet/runtime/blob/9ceb2e06dd201058376d703c3822ef56d2827977/src/libraries/System.Private.CoreLib/src/System/HashCode.cs#L304) because `int` can't be `null`. [Here's a demo of what the JIT spits out.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlxAMyUATOS7kA3svJHKG4inIAVWRlEYorJgHMAFNQoTsghjACU+w8aBAIIAJiHOHl6+ikqBAL4BRokmlOZWuBgAkiyuVO6e3n4GsYHGoeGR3jQA4jAYABJ4ABY8ITDOPj4x8YHJyYHqqeTlADwWAHzOFuSVvv3GxaVLw2ERBTAA/DV1jbgtEG0d5Bsb5KRd80YJJUk3pQDaALJ1TQeZ/ILOzxivIe98ggA8nw2BAmLgaAA5CDZQQOByOHwAXUu5D49g8GBgQ3KzgcGBm6x8qMWy2ucSAA)